File: akka-http-core/src/main/scala/akka/http/scaladsl/model/FormData.scala
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.http.scaladsl.model

import akka.http.impl.model.parser.CharacterClasses
import akka.http.impl.util.StringRendering
import akka.http.scaladsl.model.MediaTypes._

/**
 * Simple model for `application/x-www-form-urlencoded` form data.
 */
final case class FormData(fields: Uri.Query) {
  def toEntity: akka.http.scaladsl.model.RequestEntity =
    toEntity(`application/x-www-form-urlencoded`.charset)

  @deprecated("FormData always uses charset UTF-8 without appending the charset to 'Content-Type: application/x-www-form-urlencoded', use toEntity() instead.", "10.1.7")
  def toEntity(charset: HttpCharset): akka.http.scaladsl.model.RequestEntity = {
    val render: StringRendering = UriRendering.renderQuery(new StringRendering, this.fields, charset.nioCharset, CharacterClasses.unreserved)
    HttpEntity(`application/x-www-form-urlencoded`, render.get)
  }
}

object FormData {
  val Empty = FormData(Uri.Query.Empty)

  def apply(fields: Map[String, String]): FormData =
    if (fields.isEmpty) Empty else FormData(Uri.Query(fields))

  def apply(fields: (String, String)*): FormData =
    if (fields.isEmpty) Empty else FormData(Uri.Query(fields: _*))
}

File: akka-http-core/src/main/java/akka/http/javadsl/model/FormData.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.http.javadsl.model;

import java.util.Map;

import akka.japi.Pair;

/**
 * Simple model for `application/x-www-form-urlencoded` form data.
 */
public final class FormData {

  private final Query fields;

  public FormData(Query fields) {
    this.fields = fields;
  }

  /**
   * Converts this FormData to a RequestEntity using UTF8 encoding.
   */
  public RequestEntity toEntity() {
    return toEntity(HttpCharsets.UTF_8);
  }

  /**
   * Converts this FormData to a RequestEntity using the given encoding.
   *
   * @deprecated FormData always uses charset UTF-8 without appending the charset to
   *             'Content-Type: application/x-www-form-urlencoded', use toEntity() instead.
   */
  @Deprecated
  public RequestEntity toEntity(HttpCharset charset) {
    return HttpEntities.create(ContentTypes.APPLICATION_X_WWW_FORM_URLENCODED, fields.render(charset));
  }

  /**
   * Returns empty FormData.
   */
  public static final FormData EMPTY = new FormData(Query.EMPTY);

  /**
   * Creates the FormData from the given parameters.
   */
  @SafeVarargs
  public static FormData create(Pair<String, String>... params) {
    return new FormData(Query.create(params));
  }

  /**
   * Creates the FormData from the given parameters.
   */
  public static FormData create(Map<String, String> params) {
    return new FormData(Query.create(params));
  }

  /**
   * Creates a FormData from the given parameters.
   */
  public static FormData create(Iterable<Pair<String, String>> params) {
    return new FormData(Query.create(params));
  }
}
